oauthenticator/mediawiki.py

```diff
@@ -73,61 +67,53 @@ def get(self):
 
 
-class MWOAuthHandler(BaseHandler):
-    @property
-    def executor(self):
-        if hasattr(self, '_executor'):
-            return self._executor
-        else:
-            self._executor = ThreadPoolExecutor(max_workers=12)
-            return self._executor
+class MWOAuthenticator(OAuthenticator):
+    login_service = 'MediaWiki'
+    login_handler = MWLoginHandler
 
+    mw_index_url = Unicode(
+        os.environ.get('MW_INDEX_URL', 'https://meta.wikimedia.org/w/index.php'),
+        config=True,
+        help='Full path to index.php of the MW instance to use to log in'
+    )
+    
+    executor_threads = Integer(12,
+        help="""Number of executor threads.
+        
+        MediaWiki OAuth requests happen in this thread,
+        so it is mostly waiting for network replies.
+        """,
+        config=True,
+    )
+    executor = Any()
+    def _executor_default(self):
+        return ThreadPoolExecutor(self.executor_threads)
+    
     @gen.coroutine
-    def get(self):
+    def authenticate(self, handler):
+        code = handler.get_argument('code', False)
+        if not code:
+            raise web.HTTPError(400, "oauth callback made without a token")
+
         consumer_token = ConsumerToken(
-            self.authenticator.client_id,
-            self.authenticator.client_secret
+            self.client_id,
+            self.client_secret,
         )
 
         handshaker = Handshaker(
-            self.authenticator.mw_index_url, consumer_token
+            self.mw_index_url, consumer_token
         )
-        request_token = dejsonify(self.get_secure_cookie(AUTH_REQUEST_COOKIE_NAME))
-        self.clear_cookie(AUTH_REQUEST_COOKIE_NAME)
+        request_token = dejsonify(handler.get_secure_cookie(AUTH_REQUEST_COOKIE_NAME))
+        handler.clear_cookie(AUTH_REQUEST_COOKIE_NAME)
         access_token = yield self.executor.submit(
-            handshaker.complete, request_token, self.request.query
+            handshaker.complete, request_token, handler.request.query
         )
 
-        identity = handshaker.identify(access_token)
+        identity = yield self.executor.submit(handshaker.identify, access_token)
         if identity and 'username' in identity:
-            # FIXME: Figure out total set of chars that can be present
-            # in MW's usernames, and set of chars valid in jupyterhub
-            # usernames, and do a proper mapping
-            username = identity['username'].replace(' ', '_')
-            user = self.find_user(username)
-            if user is None:
-                user = orm.User(name=username, id=identity['sub'])
-                self.db.add(user)
-                self.db.commit()
-            self.set_login_cookie(user)
-            self.redirect(url_path_join(self.hub.server.base_url, 'home'))
+            # this shouldn't be necessary anymore,
+            # but keep for backward-compatibility
+            return identity['username'].replace(' ', '_')
```

Comment:
I don't entirely remember, but MW is... very weird with usernames. The user Yuvipanda is the same as the user yuvipanda (first character of usernames is case insensitive) but different from YuviPanda (rest of characters in usernames are case sensitive!). Also user named 'Some User' is equivalent to a user named 'Some_User' and 'some_User'... 

There's theoretically a canonical form, but I've had it come back with spaces instead of _s sometimes, and it's confusing. What this really needs is for me to look at the username canonicalization code in the MW php and port it here. Not a requirement for here tho :D